model.ipynb

```diff
@@ -202,30 +237,122 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "#### Selecting and training a model"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Currently, features are not scaled"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# NOTE: Running this cell can take a while!\n",
+    "# Pipeline: SelectKBest and RandomForest\n",
+    "# SelectKBest is used for features reduction\n",
+    "selectbest = SelectKBest(score_func=mutual_info_regression, k='all')\n",
+    "forest = RandomForestRegressor(criterion='mse')\n",
+    "# Creating a pipeline\n",
+    "pipe = Pipeline(steps=[('kbest', selectbest), ('forest', forest)])\n",
+    "# Parameters of pipeline for the randomized search with cross-validation\n",
+    "param_dists = {'kbest__k': stats.randint(low=32, high=128), \n",
+    "               'forest__n_estimators': stats.randint(low=200, high=1000),\n",
+    "               'forest__max_depth': [1, 3, 5, None]}\n",
+    "NITER = 100  # number of random search iterations\n",
     "time_start = timeit.default_timer()\n",
-    "parameters = {'n_estimators': [100, 500, 1000], 'max_depth': [1, 3, None]}\n",
-    "rfr_gs = GridSearchCV(estimator=RandomForestRegressor(criterion='mse'),\n",
-    "                     param_grid=parameters, cv=TimeSeriesSplit(n_splits=3),\n",
-    "                     n_jobs=-1)\n",
-    "rfr_gs.fit(X_train, y_train)\n",
-    "print(rfr_gs.best_params_)\n",
+    "search = RandomizedSearchCV(estimator=pipe, param_distributions=param_dists, \n",
+    "                            cv=TimeSeriesSplit(n_splits=3),\n",
+    "                            scoring='neg_mean_squared_error',\n",
+    "                            n_iter=NITER, refit=True, n_jobs=-1)\n",
+    "search.fit(X_train, y_train)\n",
     "time_end = timeit.default_timer()\n",
     "time_elapsed = time_end - time_start\n",
-    "print('Execution time (hour:min:sec): {}'.format(str(dt.timedelta(seconds=time_elapsed))))"
+    "print('Execution time (hour:min:sec): {}'.format(str(dt.timedelta(seconds=time_elapsed))))\n",
+    "print('Best parameter (CV score = {:.3f}):'.format(search.best_score_))\n",
+    "print(search.best_params_)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "X_train.shape"
    ]
   },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": []
+  },
   {
    "cell_type": "code",
    "execution_count": null,
    "metadata": {},
    "outputs": [],
    "source": [
-    "# Choosing optimal hyperparameters from Grid Search output\n",
-    "rfr = RandomForestRegressor(criterion='mse', **rfr_gs.best_params_)\n",
+    "# Choosing optimal hyperparameters from RandomizedSearchCV output\n",
+    "rfr = RandomForestRegressor(criterion='mse', max_depth=None, n_estimators=687)\n",
```

Comment:
Ovo se ne radi ručno! `rfr = RandomForestRegressor(criterion='mse', **rfr_gs.best_params_)` Ovaj dio sa "**" služi kao tzv. argument unpacking. Sve treba biti automatizirano! Osim toga, korišten je argument `refit=True` kod `RandomizedSearchCV` što znači da `fit` metoda "vrača" najbolji regresor. Pogledajte malo detaljnije kako se koristi `scikit-learn` API. Konačno, `Pipeline` povezuje fazu selekcije najkorisnijih featuresa i treniranje regresora (featuresi će najvjerojatnije biti različiti kako se bude mjenjao vremenski horizont i cijeli sustav stoga mora biti potpuno automatiziran).